### PR TITLE
replace builder macros with a custom scout builder

### PR DESCRIPTION
**Pull request by a-drew**
## Summary
This PR swaps out Scout Builder macro calls in the Service Provider with a custom scout builder implementation. 

This serves to clean up the service provider but also improves type hinting in IDEs. Furthermore, it lets package users swap in their own Builder with any additional changes they may need.

# Testing
- Attempt to use any of the explorer specific methods (eg: must / should / filter) covered in the BuiderTest file

### CODE REVIEW

**Comment by a-drew** [comment]
@blackshadev / @Jeroen-G any tips for passing the mutation tests?

**Comment by blackshadev** [comment]
> @blackshadev / @Jeroen-G any tips for passing the mutation tests?

I would start by writing an actual unit test for your builder. Infection will then steer you in the direction to test every edge case and method, also your builder isn't really dependent from laravel, which is great, that makes testing easier.

So create a file BuilderTest in the unit test directory. Add a test case per method to check if it actually mutates the builder. And add a testcase for the "empty" state, (so without any methods called on Builder). 

When you have done that you should be finished or close to, if you submit the code Infection will give you the lines of code which are : Uncovered, or cases with lines and mutations which didn't pass the mutation test.

Is this the information you were looking for?

**Comment by a-drew** [comment]
@blackshadev thanks for your help. After adding some proper testing, the mutation tests aren't failing anymore. 

This PR's also ready for your guys to review.

File: src/Infrastructure/Scout/Builder.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+declare(strict_types=1);
+
+namespace JeroenG\Explorer\Infrastructure\Scout;
+
+use JeroenG\Explorer\Application\Paginator;
+use JeroenG\Explorer\Domain\Aggregations\AggregationSyntaxInterface;
+use JeroenG\Explorer\Domain\Query\QueryProperties\QueryProperty;
+use JeroenG\Explorer\Domain\Syntax\Compound\BoolQuery;
+
+class Builder extends \Laravel\Scout\Builder
```
Comment (Jeroen-G) [inline src/Infrastructure/Scout/Builder.php:12]: Hmm. I would like to see this either marked as a final class. I don't want to maintain this as a public API in the package.

**Review by Jeroen-G — CHANGES_REQUESTED** [review]
Awesome idea, also makes it future proof if PHP drops dynamic properties

File: src/Infrastructure/Scout/Builder.php (hunk 1 above)
Comment (a-drew) [reply src/Infrastructure/Scout/Builder.php:12]: ahhh shame, this could be a bit of a issue on my side. As a package user, I'm extending Builder as well to add the join feature and override pagination behaviour. If the class becomes final I would need to duplicate or decorate this Builder and rewrite any of the existing methods. 

Any ideas that could work for the both of us? Ultimately this is your package so I don't want to add anything that goes against your vision.

Comment (Jeroen-G) [reply src/Infrastructure/Scout/Builder.php:12]: https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/internal.html

This might be a compromise, it'll make extending it your own risk. :)

Comment (a-drew) [reply src/Infrastructure/Scout/Builder.php:15]: For the time being, I've added the internal class tag but I do like the idea of extending a CommandBuilder and then in other cases defaulting back to the scout wrapper. 

I know that `elastic-scout-driver-plus` handles it by adding a searchQuery() method to Searchable and returns a common SearchRequestBuilder class for any advanced querying. This way there's a clear distinction between typical scout behaviour and any advanced use cases.

**Comment by a-drew** [comment]
Could use some help debugging with this phpstan error from the last CI run:
`Call to an undefined method Illuminate\Database\Eloquent\Model::getScoutKey()`

I'll try running `ci.yml` with [act](https://github.com/nektos/act) at some point, see if I can figure out why this is suddenly happening.

**Comment by blackshadev** [comment]
You can run phpstan locally with ` vendor/bin/phpstan analyze`

**Comment by blackshadev** [comment]
> Could use some help debugging this phpstan error from the [last CI run](https://github.com/Jeroen-G/Explorer/actions/runs/4693632115/jobs/8320850633): `Call to an undefined method Illuminate\Database\Eloquent\Model::getScoutKey()`
> 
> I'll try running `ci.yml` with [act](https://github.com/nektos/act) at some point, see if I can figure out why this is suddenly happening.

After some digging I found a solution. Due to Scout being updated and finally having more typing it spotted a potential problem. 
If you change the doc block in `ElasticEngine.php:66` to ` \Illuminate\Database\Eloquent\Collection<array-key, Model&Explored>` it should work again. 
Alternatively you can add an assert into line 81:
` Assert::isInstanceOf($model, Explored::class);` (above the `$this->documentAdapter->delete....`)

**Comment by a-drew** [comment]
interesting, in my case running phpstan on local always passed, even without the fix you found. that's why I wanted to test with the gh actions runner.

**Comment by a-drew** [comment]
thank you @blackshadev! all green now 🟢

**Comment by blackshadev** [comment]
Merged! Jeroen mentioned he will tag a release this weekend
